workflow/controller/operator.go

```diff
@@ -486,51 +477,43 @@ func (woc *wfOperationCtx) podReconciliation() error {
 			if node.Completed() && !node.IsDaemoned() {
 				if tmpVal, tmpOk := pod.Labels[common.LabelKeyCompleted]; tmpOk {
 					if tmpVal == "true" {
-						return nodeID
+						return
 					}
 				}
 				woc.completedPods[pod.ObjectMeta.Name] = true
 			}
 		}
-		return nodeID
+		return
 	}
 
 	parallelPodNum := make(chan string, 500)
 	var wg sync.WaitGroup
-
-	woc.currentWFSize = woc.getSize()
+	origNodeStatus := *woc.wf.Status.DeepCopy()
 
 	for _, pod := range podList.Items {
 		parallelPodNum <- pod.Name
 		wg.Add(1)
 		go func(tmpPod apiv1.Pod) {
 			defer wg.Done()
-			wfNodesLock.Lock()
-			origNodeStatus := *woc.wf.Status.DeepCopy()
-			wfNodesLock.Unlock()
-			nodeID := performAssessment(&tmpPod)
+			performAssessment(&tmpPod)
 			err = woc.applyExecutionControl(&tmpPod, wfNodesLock)
 			if err != nil {
 				woc.log.Warnf("Failed to apply execution control to pod %s", tmpPod.Name)
 			}
-			wfNodesLock.Lock()
-			defer wfNodesLock.Unlock()
-			err = woc.checkAndEstimate(nodeID)
-			if err != nil {
-				woc.wf.Status = origNodeStatus
-				nodeNameForPod := tmpPod.Annotations[common.AnnotationKeyNodeName]
-				woc.log.Warnf("%v", err)
-				woc.markNodeErrorClearOuput(nodeNameForPod, err)
-				err = woc.checkAndCompress()
-				if err != nil {
-					woc.isWFCompressionFailed = true
-				}
-			}
 			<-parallelPodNum
 		}(pod)
 	}
 
 	wg.Wait()
+
+	err = woc.checkAndCompress()
```

Comment:
if you don't need to check the size for each node,  you can move the checkandcompress from here to persistUpdates. 
persistUpdates has checkAndCompress. you need to include below code 
'''	if err != nil {
		woc.wf.Status = origNodeStatus
		woc.log.Warnf("%v", err)
		err = woc.checkAndCompress()
		woc.markWorkflowFailed(err.Error())
	}
'''

Reply:
Discussed with @jessesue. We will go with workflow failure instead of node level check.  I am looking at the Argo data persistence feature for 2.4 milestones. We don't need to worry about size, once the persistence feature implemented

Reply:
awesome